### ISSUE TITLE
Sysext images with mismatching partition architecture type get loaded

### ISSUE THREAD

**Issue by pothos** [systemd/systemd#24318]
### systemd version the issue has been seen with

251

### Used distribution

Fedora mkosi nspawn

### CPU architectures issue was seen on

x86_64

### Component

systemd-sysext

### Expected behaviour you didn't see

Following https://systemd.io/DISCOVERABLE_PARTITIONS/ a GPT sysext image containing a partition with a GUID referring to an architecture that mismatches the host should not be loaded.

### Unexpected behaviour you saw

A sysext image with a `Root Partition (64-bit ARM/AArch64)` type gets loaded on an amd64 host.

To resolve this I suggest that systemd-sysext skips over GUIDs that indicate a mismatching architecture. The usual `Linux filesystem (0FC63DAF-8483-4772-8E79-3D69D8477DE4)` or other one should not be skipped over because it has no architecture indicated. Only those GPT partition types in https://systemd.io/DISCOVERABLE_PARTITIONS/ that are defined for other architectures than the host has should be skipped.

### Steps to reproduce the problem

Create a sysext that gets loaded on your host when it's a directory.
Make a GPT disk image with one partition having GUID `b921b045-1df0-41c3-af44-4c6f280d3fae` (e.g., use `cfdisk`) and a ext4 filesystem, copy the sysext contents over.
Load it on an amd64 host.

**Comment by pothos** [comment]
I tested some more cases: With both arm64 and amd64 GUIDs present it works as expected but as soon as, e.g., an arm64 and a generic `Linux filesystem (0FC63DAF-8483-4772-8E79-3D69D8477DE4)` are present, the arm64 partition gets selected while I had expected the `Linux filesystem` to be selected.

**Comment by poettering** [comment]
ah, i see.

So the logic is that we first go by per-arch partition type uuid. And if we don't find anything we then go for a generic linux fs, but only if there's exactly one of them.

**Comment by pothos** [comment]
The case hit here is `No root partition found of the native architecture or the secondary architecture, falling back to a root partition of a non-native architecture (arm64).` (debug log) - while this look intentional this maybe is for another use case and we could opt-out of this for extension loading?

**Comment by poettering** [comment]
Hmm, is this still reproducible post 22e932f4d15f59ae773370a41a1c01ec93f8915a?

I think that dropped the whole logic, no?

**Comment by pothos** [comment]
Yes, I can reproduce it with `main` (986a514bab7f3b4c171d894804c9e5c4393886fe)
